File: final_projects/ROS1_to_ROS2_Migration_Project/jawaban_project_ros1_noetic/src/turtlesim_project_py/scripts/turtle_controller.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
import math
from my_robot_msgs.msg import Coordinates2D
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose


class TurtleController(Node):

    def __init__(self):
        super().__init__('turtle_controller')
        self.pose_ = None
        self.target_ = None
        self.cmd_vel_publisher_ = self.create_publisher(
            Twist, "/turtle1/cmd_vel", 10)
        self.pose_subscriber_ = self.create_subscription(
            Pose, "/turtle1/pose", self.callback_turtle_pose, 10)
        self.target_subscriber_ = self.create_subscription(
            Coordinates2D, "target_coordinates", self.callback_target, 10)
        self.control_loop_timer_ = self.create_timer(0.01, self.control_loop)
        self.get_logger().info("Turtle controller has been started.")

    def callback_turtle_pose(self, msg):
        self.pose_ = msg

    def callback_target(self, msg):
        if msg.x > 0.0 and msg.x < 11.0 and msg.y > 0.0 and msg.y < 11.0:
            self.get_logger().info("Received new valid target")
            self.target_ = msg

    def control_loop(self):
        if self.pose_ is None or self.target_ is None:
            return

        dist_x = self.target_.x - self.pose_.x
        dist_y = self.target_.y - self.pose_.y
        distance = math.sqrt(dist_x * dist_x + dist_y * dist_y)

        msg = Twist()

        if distance > 0.5:
            # position
            msg.linear.x = 2*distance

            # orientation
            goal_theta = math.atan2(dist_y, dist_x)
            diff = goal_theta - self.pose_.theta
            if diff > math.pi:
                diff -= 2*math.pi
            elif diff < -math.pi:
                diff += 2*math.pi

            msg.angular.z = 6*diff
        else:
            # target reached!
            msg.linear.x = 0.0
            msg.angular.z = 0.0
            self.target_ = None

        self.cmd_vel_publisher_.publish(msg)

def main(args=None):
    rclpy.init(args=args)
    node = TurtleController()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: final_projects/ROS1_to_ROS2_Migration_Project/jawaban_project_ros1_noetic/src/turtlesim_project_cpp/src/turtle_controller.cpp
#include <rclcpp/rclcpp.hpp>
#include <my_robot_msgs/msg/coordinates2_d.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <turtlesim/msg/pose.hpp>

class TurtleController : public rclcpp::Node
{
public:
    TurtleController() : Node("turtle_controller"), turtlesim_up_(false), target_up_(false)
    {
        cmd_vel_publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("/turtle1/cmd_vel", 10);
        pose_subscriber_ = this->create_subscription<turtlesim::msg::Pose>("/turtle1/pose", 10, std::bind(&TurtleController::callbackTurtlePose, this, std::placeholders::_1));
        target_subscriber_ = this->create_subscription<my_robot_msgs::msg::Coordinates2D>("target_coordinates", 10, std::bind(&TurtleController::callbackTarget, this, std::placeholders::_1));
        control_loop_timer_ = this->create_wall_timer(std::chrono::milliseconds(10), std::bind(&TurtleController::controlLoop, this));
        RCLCPP_INFO(this->get_logger(), "Turtle controller has been started.");
    }

private:
    turtlesim::msg::Pose pose_;
    my_robot_msgs::msg::Coordinates2D target_;
    bool turtlesim_up_;
    bool target_up_;

    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_publisher_;
    rclcpp::Subscription<turtlesim::msg::Pose>::SharedPtr pose_subscriber_;
    rclcpp::Subscription<my_robot_msgs::msg::Coordinates2D>::SharedPtr target_subscriber_;
    rclcpp::TimerBase::SharedPtr control_loop_timer_;

    void callbackTurtlePose(const turtlesim::msg::Pose::SharedPtr msg)
    {
        pose_ = *msg;
        turtlesim_up_ = true;
    }

    void callbackTarget(const my_robot_msgs::msg::Coordinates2D::SharedPtr msg)
    {
        if (msg->x > 0.0 && msg->x < 11.0 && msg->y > 0.0 && msg->y < 11.0)
        {
            RCLCPP_INFO(this->get_logger(), "Received new valid target");
            target_ = *msg;
            target_up_ = true;
        }
    }

    void controlLoop()
    {
        if (!turtlesim_up_ || !target_up_)
        {
            return;
        }

        double dist_x = target_.x - pose_.x;
        double dist_y = target_.y - pose_.y;
        double distance = std::sqrt(dist_x * dist_x + dist_y * dist_y);

        geometry_msgs::msg::Twist msg;

        if (distance > 0.5)
        {
            // position
            msg.linear.x = 2 * distance;

            // orientation
            double steering_angle = std::atan2(dist_y, dist_x);
            double angle_diff = steering_angle - pose_.theta;
            if (angle_diff > M_PI)
            {
                angle_diff -= 2 * M_PI;
            }
            else if (angle_diff < -M_PI)
            {
                angle_diff += 2 * M_PI;
            }
            msg.angular.z = 6 * angle_diff;
        }
        else
        {
            // target reached!
            msg.linear.x = 0.0;
            msg.angular.z = 0.0;
        }

        cmd_vel_publisher_->publish(msg);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TurtleController>());
    rclcpp::shutdown();
    return 0;
}